service/debugger/debugger.go

```diff
@@ -1475,6 +1487,158 @@ func (d *Debugger) Functions(filter string) ([]string, error) {
 	return funcs, nil
 }
 
+type TraceFunc struct {
+	// Function to be traced
+	Func     *proc.Function
+	// To keep track of depth in callgraph relative to root function
+	Depth    int
+	// To guard against adding duplicates to list of functions to be traced
+	visited  bool
+	// To avoid disassembling function multiple times
+	dis      bool
+	// To quickly access child functions
+	children []*proc.Function
+}
+type TraceFuncptr *TraceFunc
+
+var TraceMap = make(map[string]TraceFuncptr)
+
+func filter(fname string) bool {
+	idx := strings.Index(fname, "runtime.")
+	idx2 := strings.Index(fname, "runtime.defer")
+	idx3 := strings.Index(fname, "runtime.gorecover")
+	idx4 := strings.Index(fname, "runtime.gopanic") 
+	// Avoid runtime/internal functions to figure in the trace function list as they dont have a stack associated
+	idx5 := strings.Index(fname, "runtime/internal")
+	if (idx != -1 || idx5 != -1) && idx2 == -1 && idx3 == -1 && idx4 == -1 {
+		// Except certain special functions do not traverse runtime.* functions
+		return false
+	}
+	return true
+}
+
+func dequeue(queue *list.List) TraceFuncptr {
+	element := queue.Front()
+	queue.Remove(element)
+	data := element.Value
+	F, ok := data.(TraceFuncptr)
+	if !ok {
+		fmt.Printf("Type error\n")
+		os.Exit(1)
+	}
+	return F
+}
+
+func traverse(t proc.ValidTargets, f *proc.Function, depth int, FollowCalls int) ([]string, error) {
+
+	var queue *list.List
+	queue = list.New()
+	queue.Init()
+
+	funcs := []string{}
+
+	rootnode := &TraceFunc{Func: new(proc.Function), Depth: depth, children: []*proc.Function{}, visited: false, dis: false}
+	rootnode.Func = f
+
+	TraceMap[f.Name] = rootnode
+	queue.PushBack(TraceMap[f.Name])
+	for queue.Len() > 0 {
+		parent := dequeue(queue)
+		if parent.Depth > FollowCalls {
+			continue
+		}
+		if !parent.visited {
+			// Guard to avoid adding duplicates to funcs
+			funcs = append(funcs, parent.Func.Name)
+			parent.visited = true
+		}
+		if parent.Depth+1 > FollowCalls {
+			// Skip disassembling and computing children if we are going to cross FollowCalls
+			continue
+		}
+		deferdepth := 0
+		var deferparent TraceFuncptr = nil
+		if !parent.dis {
+			f := parent.Func
+			text, err := proc.Disassemble(t.Memory(), nil, t.Breakpoints(), t.BinInfo(), f.Entry, f.End)
+			if err != nil {
+				return nil, fmt.Errorf("disassemble failed with error %w", err)
+			}
+			for _, instr := range text {
+				if instr.IsCall() && instr.DestLoc != nil && instr.DestLoc.Fn != nil {
+					cf := instr.DestLoc.Fn
+					if !filter(cf.Name) {
+						continue
+					}
+					childnode := TraceMap[cf.Name]
+					if childnode == nil {
+						childnode = &TraceFunc{Func: new(proc.Function), Depth: parent.Depth + 1, children: []*proc.Function{}, visited: false, dis: false}
+						childnode.Func = cf
+						TraceMap[cf.Name] = childnode
+						parent.children = append(parent.children, cf)
+					} else {
```

Comment:
I think we can drop this else clause right? Since we're doing a BFS now, if we've seen the function before then we won't find anything new if we see it again at a larger depth. I think you can drop the loop below too (`for _, cf := range parent.children {`), we can just directly enqueue any new function right here, right?